rust/datafusion/src/physical_plan/type_coercion.rs

```diff
@@ -168,20 +168,35 @@ fn maybe_data_types(
 pub fn can_coerce_from(type_into: &DataType, type_from: &DataType) -> bool {
     use self::DataType::*;
     match type_into {
-        Int8 => matches!(type_from, Int8),
-        Int16 => matches!(type_from, Int8 | Int16 | UInt8),
-        Int32 => matches!(type_from, Int8 | Int16 | Int32 | UInt8 | UInt16),
+        Int8 => matches!(type_from, Int8 | Utf8 | LargeUtf8),
```

Comment:
This code seems inconsistent with the comments of `can_coerce_from`: 
```
/// Return true if a value of type `type_from` can be coerced
/// (losslessly converted) into a value of `type_to`
```

What would happen if we tried to coerce `"foo"` into an `Int8`? I think the correct answer is that that we should get a runtime error. I worry that we would end up silently converting to `Null`. 

Can you explain why these coercion rules are needed?  And if we decide they are, can we please update the comments of this function to reflect the newfound understanding?

@jorgecarleitao  I think did some previous work in this area, so he may have some relevant perspective. cc @andygrove  in case you are interested

Reply:
@alamb I think you have hit on a bigger issue.

Postgres will do this type coercion silently: `SELECT LEFT('abcde', '1');` will return `a`.  And `SELECT LEFT('abcde', 'a');` will return `invalid input syntax for type integer: "a"`.

I think the default return of a failed `CAST` in DataFusion is currently `NULL` which is not good and not an ANSI expected behavior. I will volunteer to fix it if we can reach a consensus.

This is one of my major issues with Spark. Some other engines explicitly call out this behavior with `SAFE CAST`.

Reply:
I think the cast in DataFusion relies on the `cast` in Arrow. I think Arrow should have an option (I think there is a JIRA issue for that already) that would return an error on a failing cast instead of nulls, which is I think what we then can/should use for DataFusion.

Reply:
@Dandandan @alamb if you agree I can try to add it as I still have the function work in my head.

Reply:
I agree that adding an option to the arrow cast kernel to error (rather than silently return `Null` would be very valuable). I can see usecases for returning `Null` which is why I don't think we could just change the cast kernel's behavior. But an option would be lovely ❤️ 

Sorry for the delay in responding

Reply:
No problem @alamb. How about I revert the coercion logic from this PR and re-add the explicit cast in the tests. I think there is a _major_ piece of work to fully address CAST/coercion.

Reply:
> How about I revert the coercion logic from this PR and re-add the explicit cast in the tests. 
Sounds like a great plan

> I think there is a major piece of work to fully address CAST/coercion.
I agree 💯

Reply:
Another option is to add a new function `maybe_cast -> Result<Arc<dyn Array>>`.

Reply:
@jorgecarleitao @alamb 

I would like to get the rest of these SQL functions implemented then pick up the type-coercion work. I was also thinking a `try_cast` implementation but it will be a lot of work. 

I think we also need to produce a matrix like this: https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#implicit-conversions to work out the scope of the changes as I think we have a very imbalanced set of implementations at the moment.